Add IntermediateOutputPath (#4417)
MicrosoftDocs/visualstudio-docs: docs/test/live-unit-testing-faq.md

</Target>
```

## Error messages with \<OutputPath>, \<OutDir> or \<IntermediateOutputPath>

**Why do I get the following error when Live Unit Testing tries to build my solution: "...appears to unconditionally set `<OutputPath>` or `<OutDir>`. Live Unit Testing will not execute tests from the output assembly"?**

You can get this error if the build process for your solution has custom logic that specifies where binaries should be generated. By default the location of your binaries depends on `<OutputPath>`, `<OutDir>` or `<IntermediateOutputPath>` as well as `<BaseOutputPath>` or `<BaseIntermediateOutputPath>`.

Live Unit Testing overrides those variables to ensure that build artifacts are dropped to a Live Unit Testing artifacts folder and will fail if your build process also overrides these variables.

> There are a two main approaches to make Live Unit Testing build successfully. For easier build configurations you can base your output paths on `<BaseIntermediateOutputPath>`. For more complex configurations you can base your output paths on `<LiveUnitTestingBuildRootPath>`.

Reviewer: ```suggestion
There are a two main approaches to make Live Unit Testing build successfully. For easier build configurations, you can base your output paths on `<BaseIntermediateOutputPath>`. For more complex configurations, you can base your output paths on `<LiveUnitTestingBuildRootPath>`.
```

### Overriding `<OutputPath>`/`<IntermediateOutputPath>` conditionally based on `<BaseOutputPath>`/ `<BaseIntermediateOutputPath>`.

> > Note: To use this approach each project needs to be able to build independently from one another. Do not have one project reference artifacts of another project during build. Do not have one project dynamically load assemblies from another project during runtime (for example call `Assembly.Loadfile("..\..\Project2\Release\Project2.dll")`).

Reviewer: ```suggestion
> [!NOTE]
> To use this approach each project needs to be able to build independently from one another. Do not have one project reference artifacts of another project during build. Do not have one project dynamically load assemblies from another project during runtime (for example call `Assembly.Loadfile("..\..\Project2\Release\Project2.dll")`).
```

> During build Live Unit Testing will automatically override the `<BaseOutputPath>`/`<BaseIntermediateOutputPath>` variables to target the Live Unit Testing artifacts folder.

Reviewer: ```suggestion
During build, Live Unit Testing automatically overrides the `<BaseOutputPath>`/`<BaseIntermediateOutputPath>` variables to target the Live Unit Testing artifacts folder.
```

Do not override `<OutDir>` directly in your build process; override `<OutputPath>` instead to drop build artifacts to a specific location.

### Overriding your properties based on the `<LiveUnitTestingBuildRootPath>` property.

> > Note: In this approach you need to be careful about files added under the artifacts folder that are not generated during build. The example below shows what to do with placing the packages folder under artifacts. Since the contents of this folder are not generated during the build the MSBuild property **should not be changed**.

Reviewer: ```suggestion
> [!NOTE]
> In this approach, you need to be careful about files added under the artifacts folder that are not generated during build. The example below shows what to do when placing the packages folder under artifacts. Because the contents of this folder are not generated during the build, the MSBuild property **should not be changed**.
```

> During a Live Unit Testing build the `<LiveUnitTestingBuildRootPath>` property will be set to the location of Live Unit Testing artifacts folder.

Reviewer: ```suggestion
During a Live Unit Testing build, the `<LiveUnitTestingBuildRootPath>` property is set to the location of Live Unit Testing artifacts folder.
```

> For example assume that your project has the structure as shown below.

Reviewer: ```suggestion
For example, assume that your project has the structure as shown below.
```

```
.vs\...\lut\0\b
artifacts\{binlog,obj,bin,nupkg,testresults,packages}
src\{proj1,proj2,proj3}
tests\{testproj1,testproj2}
Solution.sln
```

> During the Live Unit Testing build, the `<LiveUnitTestingBuildRootPath>` property will be set to the full path of `.vs\...\lut\0\b`. If the project defines `<ArtifactsRoot>` property that maps to the solution dir, you can update the MSBuild project as follows:

Reviewer: ```suggestion
During the Live Unit Testing build, the `<LiveUnitTestingBuildRootPath>` property is set to the full path of `.vs\...\lut\0\b`. If the project defines `<ArtifactsRoot>` property that maps to the solution dir, you can update the MSBuild project as follows:
```
